moz_security/modules/iputils.lua

```diff
@@ -0,0 +1,178 @@
+--[[
+# IP / CIDR Utility Module
+
+iputils can be used for verifying if an IP address is part of a specific set
+of CIDR / subnets.
+
+This module is based on lua-resty-iputils, which can be found at
+https://github.com/hamishforbes/lua-resty-iputils/.
+
+## Functions
+
+### parse_cidr
+
+Parse an array of CIDR subnet specifications where each entry is a string.
+Returns a list of the subnets in binary form for later use in ip_in_cidrs.
+
+If a subnet mask is not specified with a subnet entry, it will default to 32.
+
+```lua
+local ipu = require("iputils")
+
+local subnets = ipu.parse_cidrs({ "10.0.0.0/24", "192.168.1.0/24" })
+```
+
+*Arguments*
+- cidrs (table) - Array of subnets to parse, each in string format
+
+*Return*
+- bincidrs (table) - Parsed subnets in binary format
+
+### ip_in_cidrs
+
+Returns true if an IP address is in a list of binary CIDR representations
+as returned by parse_cidrs. Returns false if not present, or nil if an
+error occurs.
+
+```lua
+local ipu = require("iputils")
+
+local subnets = ipu.parse_cidrs({ "10.0.0.0/24", "192.168.1.0/24" })
+inlist = ipu.ip_in_cidrs("10.0.0.1", subnets)
+```
+
+*Arguments*
+- ip (string) - IP address to verify
+- bincidrs (table) - Parsed subnets returned from parse_cidrs
+
+*Returns*
+- flag (boolean) - True if in subnets, false if not, nil if error
+
+--]]
+
+local ipairs    = ipairs
+local tonumber  = tonumber
+local tostring  = tostring
+local type      = type
+
+local ipa       = require("lpeg.ip_address")
+local bit       = require("bit")
+local string    = require("string")
+
+local M = {}
+setfenv(1, M)
+
+
+-- Precompute binary subnet masks
+local bin_masks = {}
+for i=0,32 do
+    bin_masks[tostring(i)] = bit.lshift((2^i)-1, 32-i)
+end
+
+
+-- Precompute inverted binary subnet masks
+local bin_inverted_masks = {}
+for i=0,32 do
+    local i = tostring(i)
+    bin_inverted_masks[i] = bit.bxor(bin_masks[i], bin_masks["32"])
+end
+
+
+local function split_octets(input)
+    return ipa.v4_octets:match(input)
+end
+
+
+local function unsign(bin)
+    if bin < 0 then
+        return 4294967296 + bin
+    end
+    return bin
+end
+
+
+local function ip2bin(ip)
+    if type(ip) ~= "string" then
+        return nil
+    end
+
+    local octets = split_octets(ip)
+    if not octets or #octets ~= 4 then
```

Comment:
Don't need the length check the grammar verified it (the grammar could actually do the complete conversion and this function could go away (optional)).